tests/rustdoc-ui/extract-doctests.rs

```diff
@@ -0,0 +1,15 @@
+// Test to ensure that it generates expected output for `--extract-doctests` command-line
+// flag.
+
+//@ compile-flags:-Z unstable-options --extract-doctests
+//@ normalize-stdout-test: "tests/rustdoc-ui" -> "$$DIR"
+//@ check-pass
+
```

Comment:
We probably want a test for how this interacts with:

- Implicit/explicit `fn main()`
- Hidding lines with `/// # use some::path;`

Reply:
True although currently it's the code as defined in the documentation and no changes on rustdoc side (which I think matches better what rust for linux wants).

Reply:
Maybe we should provide another field with "rustdoc computed code" where hidden lines and `main` function wrapping are added by rustdoc. Actually that sounds like a very good idea.

Reply:
> True although currently it's the code as defined in the documentation and no changes on rustdoc side (which I think matches better what rust for linux wants).

Currently we use both hidden lines and the `?` support, i.e. the `# Ok::<..., ...>(...)` syntax (which implies the `fn main()`, from what I understand).

Some of that post-processing may be easy to do by users, like the hidden lines I assume, but if you walk the source or IR or similar to figure out details (like the crate attributes that you move out of `main()`), then it may be harder for end users to replicate that properly without a hack. Perhaps it may be possible to export what `rustdoc` figured about the test, so that users can replicate the post-processing on their side, but customized to their environment.

Having the `rustdoc` computed code sounds fine since it may be enough for some use cases, but e.g. we currently need to convert on the fly the `.unwrap()` in the `?`-using tests into a custom `assert!`.

Reply:
So, for instance, if `rustdoc` tells us "this test uses `?`", "these are the crate attributes I would have moved", etc., then users may be able to easily and reliably construct their own wrappers and e.g. do a check instead of an `.unwrap()`.

Of course, for things like crate attributes, it may be best to remove them so that the end user can re-add them where needed, so it wouldn't be the completely "unaltered" source code. But it could be an "adapted" version. Hidden lines should probably be normal lines in that adapted version.

Perhaps it makes sense to provide all those versions, i.e. the completely unaltered one for those that may want to do something complex or to render the text somewhere, the "adapted" version for customized test environments and the `rustdoc` computed code for those that can use directly that. In the kernel we would use the "adapted" one, only, I think.